### PR TITLE
Add trace information to eth_estimateGas

### LINKED ISSUE #10360: Feature request: Include trace information when virtual calls throw exceptions
- **Parity Ethereum version**: 2.2.9-stable
- **Operating system**: Linux
- **Installation**: one-line installer
- **Fully synchronized**: yes
- **Network**: ethereum, kovan
- **Restarted**: yes

Related to https://github.com/paritytech/parity-ethereum/issues/10147. The error message thrown when `eth_estimateGas` (and I assume `eth_call` and siblings) is very cryptic and could be improved by including diagnostic info and VM trace info if applicable.

### PR DESCRIPTION
**Pull request by fanatid**
Closes #10360

### CODE REVIEW

**Comment by joshua-mir** [comment]
This should probably be in the `parity_` namespace as it breaks spec.

File: ethcore/src/client/client.rs [hunk 1]
```diff
@@ -1570,22 +1570,28 @@ impl Call for Client {
...
-		let cond = |gas| exec(gas).unwrap_or(false);
-
 		if !cond(upper) {
 			upper = max_upper;
 			match exec(upper) {
-				Some(false) => return Err(CallError::Exceptional),
-				None => {
+				Ok(v) => {
+					if v.exception.is_some() {
+						return Err(CallError::Exceptional(v.exception.unwrap()))
```
Comment (ordian) [inline ethcore/src/client/client.rs]: Could you use an `if let Some(exception) = v.exception` statement to avoid `.unwrap()`?

Comment (fanatid) [reply ethcore/src/client/client.rs]: I had read in docs that `.unwrap()` can cause panic, but before I check with `.is_some()`, so thought that this is fine.
Sorry, it's not clear for me, how `.unwrap()` can be avoided with `if let Some(exception) = v.exception`... can you explain?

Comment (ordian) [reply ethcore/src/client/client.rs]: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=132afa5ce895d6d8d46882ee4df7d904

Comment (fanatid) [reply ethcore/src/client/client.rs]: Thank you for example.

File: rpc/src/v1/helpers/errors.rs [hunk 2]
```diff
@@ -173,11 +173,11 @@ pub fn state_corrupt() -> Error {
 	internal("State corrupt", "")
 }
 
-pub fn exceptional() -> Error {
+pub fn exceptional<T: fmt::Debug>(data: T) -> Error {
```
Comment (ordian) [inline rpc/src/v1/helpers/errors.rs]: I think `fmt::Display` implementation for `vm::Error` is more user-friendly.

Comment (fanatid) [reply rpc/src/v1/helpers/errors.rs]: I'm I understand correct that if we switch to `fmt::Display` this formatter will be used?
https://github.com/paritytech/parity-ethereum/blob/9cb86061032a323710a1dd549e1e1f0a4e33eedd/ethcore/vm/src/error.rs#L100-L117

Comment (ordian) [reply rpc/src/v1/helpers/errors.rs]: Yes, that's my suggestion, but I don't have a strong opinion about that.

File: rpc/src/v1/helpers/errors.rs [hunk 3]
```diff
@@ -173,11 +173,11 @@ pub fn state_corrupt() -> Error {
 	internal("State corrupt", "")
 }
 
-pub fn exceptional() -> Error {
+pub fn exceptional<T: fmt::Display>(data: T) -> Error {
 	Error {
 		code: ErrorCode::ServerError(codes::EXCEPTION_ERROR),
 		message: "The execution failed due to an exception.".into(),
-		data: None,
+		data: Some(Value::String(format!("{:?}", data))),
```
Comment (ordian) [inline rpc/src/v1/helpers/errors.rs]:
```suggestion
		data: Some(Value::String(data.to_string())),
```

File: ethcore/src/client/client.rs [hunk 4]
```diff
@@ -1570,22 +1570,28 @@ impl Call for Client {
 			let schedule = machine.schedule(env_info.number);
 			Executive::new(&mut clone, &env_info, &machine, &schedule)
 				.transact_virtual(&tx, options())
+		};
+
+		let cond = |gas| {
+			exec(gas)
 				.ok()
 				.map(|r| r.exception.is_none())
+				.unwrap_or(false)
```
Comment (niklasad1) [inline ethcore/src/client/client.rs]: could be `map_or(false, |r| r.exception.is_none())`

**Pull request merged**